### PR TITLE
Allow testing of alertmanager gRPC replication in multitenant_test.go.

### PR DESCRIPTION
**Pull request by stevesg**
**What this PR does**:
Allow testing of alertmanager gRPC replication in multitenant_test.go.

Use a fake implementation of the alertmanager ClientPool to connect
together the alertmanager instances in the multitenant_test.go unit
test. This allows testing of the replication logic without relying on
doing do at integration test level, and without having the weight of the
gRPC server running.

**Checklist**
- [X] ~~Tests updated~~
- [X] ~~Documentation added~~
- [X] ~~`CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`~~

### CODE REVIEW

File: pkg/alertmanager/multitenant_test.go [hunk 1]
```diff
@@ -1178,13 +1188,19 @@ func TestAlertmanager_StateReplicationWithSharding(t *testing.T) {
...
-			assert.Equal(t, float64(1), metrics.GetSumOfGauges("cortex_alertmanager_silences"))
-			assert.Equal(t, float64(1), metrics.GetSumOfCounters("cortex_alertmanager_state_replication_total"))
+			assert.Equal(t, float64(tt.replicationFactor), metrics.GetSumOfGauges("cortex_alertmanager_silences"))
+			assert.Equal(t, float64(tt.replicationFactor), metrics.GetSumOfCounters("cortex_alertmanager_state_replication_total"))
 			assert.Equal(t, float64(0), metrics.GetSumOfCounters("cortex_alertmanager_state_replication_failed_total"))
 
-			// 5b. We'll never receive the message as GRPC communication over unit tests is not possible.
-			assert.Equal(t, float64(0), metrics.GetSumOfCounters("alertmanager_partial_state_merges_total"))
-			assert.Equal(t, float64(0), metrics.GetSumOfCounters("alertmanager_partial_state_merges_failed_total"))
+			// 5b. Check the number of partial states merged are as we expect.
```
[the remark this answers is not in the document]
Comment (stevesg) [reply pkg/alertmanager/multitenant_test.go:1195]: This is expected as of right now.

File: pkg/alertmanager/multitenant_test.go [hunk 2]
```diff
@@ -1178,13 +1188,19 @@ func TestAlertmanager_StateReplicationWithSharding(t *testing.T) {
...
+			// Partial states are currently replicated twice:
+			//   For RF=1 1 -> 0      = Total 0 merges
+			//   For RF=2 1 -> 1 -> 1 = Total 2 merges
+			//   For RF=3 1 -> 2 -> 4 = Total 6 merges
+			nFanOut := tt.replicationFactor - 1
+			nMerges := nFanOut + (nFanOut * nFanOut)
+			assert.Equal(t, float64(nMerges), metrics.GetSumOfCounters("cortex_alertmanager_partial_state_merges_total"))
```
[the remark this answers is not in the document]
Comment (pracucci) [reply pkg/alertmanager/multitenant_test.go:1202]: This is correct 👍

**Review by pracucci — APPROVED** [review]
Good job, LGTM!

**Review by jtlisi — APPROVED** [review]
LGTM, good work!

**Pull request merged**